src/solvers/imex_tableaus.jl

```diff
@@ -6,39 +8,91 @@ export SSP222, SSP322, SSP332, SSP333, SSP433
 
 """
-    IMEXARKTableau(; a_exp, b_exp, c_exp, a_imp, b_imp, c_imp)
+    IMEXTableau(; a_exp, b_exp, c_exp, a_imp, b_imp, c_imp)
+
+A wrapper for an IMEX Butcher tableau (or, more accurately, a pair of Butcher
+tableaus, one for explicit tendencies and the other for implicit tendencies).
+Only `a_exp` and `a_imp` are required arguments; the default values for `b_exp`
+and `b_imp` assume that the algorithm is FSAL (first same as last), and the
+default values for `c_exp` and `c_imp` assume that it is internally consistent.
 
-Generates an `IMEXARKTableau` struct from an IMEX ARK Butcher tableau. Only
-`a_exp` and `a_imp` are required arguments; the default values for `b_exp` and
-`b_imp` assume that the algorithm is FSAL (first same as last), and the default
-values for `c_exp` and `c_imp` assume that the algorithm is internally
-consistent.
+The explicit tableau must be strictly lower triangular, and the implicit tableau
+must be lower triangular (only DIRK algorithms are currently supported).
 """
-struct IMEXARKTableau{VS <: StaticArrays.StaticArray, MS <: StaticArrays.StaticArray} <: AbstractTableau
+struct IMEXTableau{VS <: StaticArrays.StaticArray, MS <: StaticArrays.StaticArray}
     a_exp::MS # matrix of size s×s
     b_exp::VS # vector of length s
     c_exp::VS # vector of length s
     a_imp::MS # matrix of size s×s
     b_imp::VS # vector of length s
     c_imp::VS # vector of length s
 end
-function IMEXARKTableau(;
+function IMEXTableau(;
     a_exp,
     b_exp = a_exp[end, :],
     c_exp = vec(sum(a_exp; dims = 2)),
     a_imp,
     b_imp = a_imp[end, :],
     c_imp = vec(sum(a_imp; dims = 2)),
 )
+    @assert all(iszero, UpperTriangular(a_exp))
+    @assert all(iszero, UpperTriangular(a_imp) - Diagonal(a_imp))
+
     # TODO: add generic promote_eltype
     a_exp, a_imp = promote(a_exp, a_imp)
     b_exp, b_imp, c_exp, c_imp = promote(b_exp, b_imp, c_exp, c_imp)
-    return IMEXARKTableau(a_exp, b_exp, c_exp, a_imp, b_imp, c_imp)
+    return IMEXTableau(a_exp, b_exp, c_exp, a_imp, b_imp, c_imp)
 end
 
+"""
+    IMEXAlgorithm(name, newtons_method, [constraint])
+    IMEXAlgorithm(tableau, newtons_method, [constraint])
+
+Runs an IMEX algorithm with a particular IMEX Butcher tableau. The first
+constructor can be used to automatically determine the tableau (and the default
+constraint) from an `AbstractAlgorithmName`, while the second constructor
+can be used to specify a custom tableau (an an optional constraint).
+"""
+struct IMEXAlgorithm{C <: AbstractAlgorithmConstraint, T <: IMEXTableau, N <: NewtonsMethod} <: DistributedODEAlgorithm
+    tableau::T
+    newtons_method::N
+    constraint::C
+end
+IMEXAlgorithm(tableau::IMEXTableau, newtons_method, constraint = Unconstrained()) =
+    IMEXAlgorithm(tableau, newtons_method, constraint)
+IMEXAlgorithm(name::AbstractAlgorithmName, newtons_method, constraint = default_constraint(name)) =
+    IMEXAlgorithm(IMEXTableau(name), newtons_method, constraint)
```

Comment:
```suggestion
struct IMEXAlgorithm{C <: AbstractAlgorithmConstraint, T <: IMEXTableau, N <: NewtonsMethod} <: DistributedODEAlgorithm
    constraint::C
    tableau::T
    newtons_method::N
end
IMEXAlgorithm(tableau::IMEXTableau, newtons_method) =
    IMEXAlgorithm(Unconstrained(), tableau, newtons_method)
IMEXAlgorithm(name::AbstractAlgorithmName, newtons_method, constraint = default_constraint(name)) =
    IMEXAlgorithm(constraint, IMEXTableau(name), newtons_method)
```
I think we need something like this instead, to work properly. (There's infinite recursion here).  Also, the type parameters are scrambled compared to the properties, so I've fixed that too. There may be other places where we call the default constructor, so those will need to be fixed, too. We could alternatively use 
```julia
struct IMEXAlgorithm{T <: IMEXTableau, N <: NewtonsMethod, C <: AbstractAlgorithmConstraint} <: DistributedODEAlgorithm
    tableau::T
    newtons_method::N
    constraint::C
end
```
And adjust how specialization works on the constraint.

Reply:
This all looks good. I'll add it in. The constraint should probably be the first parameter so that we can dispatch off of `IMEXAlgorithm{SSPConstrained}`, rather than `IMEXAlgorithm{<:Any, <:Any, SSPConstrained}`.

Reply:
Specializing is fine. We could alternatively use dispatch by passing `foo(alg, alg.constraint)` and catch via `foo(:: IMEXAlgorithm, :: SSPConstrained)` etc.